### PR TITLE
Congestion hashmap implementation

### PR DESCRIPTION
**Pull request by chaupow**
Working towards https://github.com/mapbox/directions/issues/501

### CODE REVIEW

File: src/csv.hpp [hunk 1]
```diff
@@ -0,0 +1,1248 @@
...
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+//    this list of conditions and the following disclaimer.
+//
+//2. Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
```
Comment (daniel-j-h) [inline src/csv.hpp]: ^

File: src/csv.hpp [hunk 2]
```diff
@@ -0,0 +1,1248 @@
+// Copyright: (2012-2015) Ben Strasser <code@ben-strasser.net>
```
Comment (daniel-j-h) [inline src/csv.hpp]: Is

https://github.com/Project-OSRM/osrm-backend/blob/ffd6311e7ddabcf9b6c829e7c79a6e708d66dc51/include/updater/csv_file_parser.hpp

"good enough"?

File: src/hashmap.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#include <fstream>
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: Include guards missing

File: src/hashmap.hpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#include <iostream>
+#include <sparsepp/spp.h>
+
+#include "types.hpp"
+#include <sparsepp/spp.h>
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: spp.h is included twice

File: src/hashmap.hpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        }
+    };
+}
+
+class Hashmap {
+    public:
+        Hashmap();
+        // Hashmap(std::ifstream& input);
+        Hashmap(const std::string &input_filename);
+        inline void add(const external_nodeid_t &to, const external_nodeid_t &from, const congestion_speed_t &speed);
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: Why the inline here?

File: src/hashmap.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        congestion_speed_t getValue(external_nodeid_t to, external_nodeid_t from) const;
+        bool hasKey(external_nodeid_t to, external_nodeid_t from) const;
+        std::vector<congestion_speed_t> getValues(std::vector<external_nodeid_t>& way) const;
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: From the function signature alone: I don't understand the semantics here. Why is `getValues` able to modify the `way` vector? Should it be a `const &`?

File: src/hashmap.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    private:
+        sparse_hash_map<Way, int> annotations;
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: What is the `int` value here we map to?

File: src/lookup_bindings.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      return Nan::ThrowTypeError("NodeIds must be an array of numbers");
+    auto nodeId = Nan::To<external_nodeid_t>(jsNodeId).FromJust();
+    resulting_nodeIds[i] = nodeId;
+  }
+
+  struct Worker final : Nan::AsyncWorker {
+    using Base = Nan::AsyncWorker;
+
+    Worker(std::shared_ptr<Hashmap> annotations, Nan::Callback* callback, std::vector<external_nodeid_t> nodeIds)
+        : Base(callback), annotations{std::move(annotations)}, nodeIds(nodeIds) {}
```
Comment (daniel-j-h) [inline src/lookup_bindings.cpp]: `nodeIds` is taken by value here, and then copied into `nodeIds` - I don't think this is by intention

File: src/types.hpp [hunk 9]
```diff
@@ -7,9 +7,12 @@
 
 // Type declarations for node ids and way ids.  We don't need the full 64 bits
 // for ways as the highest way number isn't close to 2^32 yet.
-typedef std::uint64_t external_nodeid_t;
+typedef std::int64_t external_nodeid_t;
```
Comment (daniel-j-h) [inline src/types.hpp]: Why signed now?

File: test/basic/rtree.cpp [hunk 10]
```diff
@@ -54,7 +54,7 @@ BOOST_AUTO_TEST_CASE(rtree_radius_test)
     BOOST_CHECK_EQUAL(results[0], INVALID_INTERNAL_NODEID);
 
     // This should be within about 1.1m
-    points = std::vector<point_t>{point_t{1.00001, 1}};
+    points = std::vector<point_t>{point_t{1.00005, 1}};
```
Comment (daniel-j-h) [inline test/basic/rtree.cpp]: Why the unrelated change here? Should the test compare with an epsilon?

File: test/congestion/congestion.csv [hunk 11]
```diff
@@ -0,0 +1,100 @@
+62397298,62523814,20
```
Comment (daniel-j-h) [inline test/congestion/congestion.csv]: Could you add tests for invalid csv files / not numbers / negative numbers / more columns / `#header` etc.

File: src/csv.hpp (hunk 2 above)
Comment (ghoshkaj) [inline src/csv.hpp]: No, it's slower than regular std::getline.

File: src/csv.hpp (hunk 1 above)
Comment (ghoshkaj) [inline src/csv.hpp]: I'm not sure what you are trying to say here.

File: test/basic/rtree.cpp (hunk 10 above)
Comment (ghoshkaj) [inline test/basic/rtree.cpp]: This was an error in master that we had to manually fix here when rebasing against Dane's build system changes.

File: src/types.hpp (hunk 9 above)
Comment (ghoshkaj) [inline src/types.hpp]: Because NaN doesn't like uint64_t:
```
/.deps/Release/obj.target/route_annotator/src/lookup_bindings.o.d.raw   -c -o Release/obj.target/route_annotator/src/lookup_bindings.o ../src/lookup_bindings.cpp
../src/lookup_bindings.cpp:68:19: error: no matching function for call to 'To'
    auto nodeId = Nan::To<external_nodeid_t>(jsNodeId).FromJust();
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/kajari/Mapbox/route-annotator/node_modules/nan/nan_converters.h:53:38: note: candidate template ignored: substitution failure [with T = unsigned long long]: implicit instantiation of undefined template 'Nan::imp::ToFactory<unsigned long long>'
typename imp::ToFactory<T>::return_t To(v8::Local<v8::Value> val) {
~~~~~~~~
```

File: src/hashmap.hpp (hunk 6 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: It shouldn't! You're right! It should be a `const &`!

File: src/hashmap.hpp (hunk 7 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: This should be `congestion_speed_t`. Thanks Daniel!

File: src/lookup_bindings.cpp (hunk 8 above)
Comment (ghoshkaj) [inline src/lookup_bindings.cpp]: @danpat explained what this means to me! Okay so I've wrapped `nodeIds(std::move(nodeIds))`

File: src/csv.hpp (hunk 2 above)
Comment (ghoshkaj) [inline src/csv.hpp]: It was actually 2 minutes slower than regular std::getline! I talked about it here: https://github.com/mapbox/directions/issues/501#issuecomment-297209471

Comment (daniel-j-h) [inline src/csv.hpp]: Conversation in osrm-backend https://github.com/Project-OSRM/osrm-backend/issues/3974

File: src/types.hpp (hunk 9 above)
Comment (daniel-j-h) [inline src/types.hpp]: Yep check
https://github.com/nodejs/nan/blob/master/doc/converters.md#api_nan_to
I would use the `To<int64_t>` function, assert for `> 0` and `static_cast` to your unsigned type then. Negative ids do not make sense after all - the Js->C++ code should assert that.

File: src/hashmap.hpp (hunk 3 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: You're right! Added them!

File: src/types.hpp (hunk 9 above)
Comment (ghoshkaj) [inline src/types.hpp]: Cool! Done! Thank you for that suggestion!

File: CMakeLists.txt [hunk 12]
```diff
@@ -45,6 +46,7 @@ IF(NODEJS)
   FIND_PACKAGE(NodeJS)
 
   ADD_NODEJS_MODULE(${PROJECT_NAME} src/nodejs_bindings.cpp)
+  ADD_NODEJS_MODULE(${PROJECT_NAME} src/lookup_bindings.cpp)
```
Comment (daniel-j-h) [inline CMakeLists.txt]: I don't understand this - should this be a single

    ADD_NODEJS_MODULE(${PROJECT_NAME} src/nodejs_bindings.cpp src/lookup_bindings.cpp)

? Not sure how node-cmake behaves if we create two node modules with the same name.

File: src/hashmap.hpp (hunk 3 above)
Comment (daniel-j-h) [inline src/hashmap.hpp]: Can you prefix them with your project's name?

The issue with generic include guard names such as `HASHMAP_H` is the probability for name clashes is high. If a header you include (or a header your included header include) defines this symbol your include guard is broken. That's the reason we use e.g. `OSRM_SOME_HEADER_NAME_H` in osrm.

File: src/hashmap.hpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            // performance.
+            return (p.to << 31) + p.from;
+        }
+    };
+}
+
+class Hashmap {
+    public:
+        Hashmap();
+        Hashmap(const std::string &input_filename);
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: Below all member functions taking a `from` and `to` id: should they take a single `Way` object instead`? From what I can see a `Way` is exactly that: a `from` and a `to` id.

File: src/hashmap.hpp [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        inline void add(const external_nodeid_t &from, const external_nodeid_t &to, const congestion_speed_t &speed);
```
Comment (daniel-j-h) [inline src/hashmap.hpp]: Remove inline

File: src/lookup_bindings.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    std::vector<external_nodeid_t> resulting_nodeIds(jsNodeIds->Length());
+
+  for (uint32_t i = 0; i < jsNodeIds->Length(); ++i)
+  {
+    v8::Local<v8::Value> jsNodeId = jsNodeIds->Get(i);
+    if (!jsNodeId->IsNumber())
+      return Nan::ThrowTypeError("NodeIds must be an array of numbers");
+    auto signedNodeId = Nan::To<int64_t>(jsNodeId).FromJust();
+    if(signedNodeId < 0)
```
Comment (daniel-j-h) [inline src/lookup_bindings.cpp]: Are we using clang-format for this project? Seems like the formatting changes from time to time.

(If not don't mind this comment)

File: src/lookup_bindings.cpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+  auto* self = new Lookup(std::move(annotations));
+
+  self->Wrap(info.This());
+  info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(Lookup::GetAnnotations) {
+  auto* const self = Nan::ObjectWrap::Unwrap<Lookup>(info.Holder());
+  (void) self;
```
Comment (daniel-j-h) [inline src/lookup_bindings.cpp]: Remove the `(void)self` - we use `self` below.

File: src/nodejs_bindings.cpp [hunk 17]
```diff
@@ -44,8 +44,8 @@ NAN_METHOD(Annotator::New)
     }
     else
     {
-        auto init = Nan::New(constructor());
-        info.GetReturnValue().Set(init->NewInstance());
+        return Nan::ThrowTypeError(
+            "Cannot call constructor as function, you need to use 'new' keyword");
```
Comment (daniel-j-h) [inline src/nodejs_bindings.cpp]: Why the change here? The `Lookup::New` function still supports both `new T()` as well as `T()`

https://github.com/mapbox/route-annotator/pull/5/files#diff-e03b7689579005881fd2f324670dbfbfR22

Either allow it both the Annotator and the Lookup Js type or neither.

File: src/hashmap.cpp [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    // Use the already retrieved value as the result
+    return result->second;
+};
+
+std::vector<congestion_speed_t> Hashmap::getValues(const std::vector<external_nodeid_t> &way) const
+{
+    std::vector<congestion_speed_t> speeds;
+    if (way.size() > 1)
```
Comment (daniel-j-h) [inline src/hashmap.cpp]: Should a way with size 0 or 1 result in an exception? This is a user-error, no?

File: src/hashmap.cpp [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+Hashmap::Hashmap() {};
+
+Hashmap::Hashmap(const std::string &input_filename) {
+
+    io::CSVReader<3> in(input_filename);
+    in.set_header("from", "to", "speed");
+    external_nodeid_t from; external_nodeid_t to; congestion_speed_t speed;
+    // Pre-allocate a large chunk of memory to save on all the micro-allocations
+    // that would happen if adding items one-by-one and growing as needed
+    annotations.reserve(139064548);
```
Comment (daniel-j-h) [inline src/hashmap.cpp]: What is this magic number? Why should the hash map reserve space / buckets for 139064548 items?

File: src/hashmap.cpp [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+inline void Hashmap::add(const external_nodeid_t &from, const external_nodeid_t &to, const congestion_speed_t &speed) {
+    annotations[Way(from,to)] = speed;
+};
+
+bool Hashmap::hasKey(external_nodeid_t from, external_nodeid_t to) const {
+    if (annotations.count(Way(from,to)) > 0) {
+        return true;
+    } else {
+        return false;
+    }
```
Comment (daniel-j-h) [inline src/hashmap.cpp]: You can always simplify

```cpp
if (p) {
  return true;
} else {
  return false;
}
```

to

```c++
return p;
```

File: src/hashmap.cpp [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+// @TODO use hasKey to get pointer and directly return
```
Comment (daniel-j-h) [inline src/hashmap.cpp]: Hm? hasKey does not return a pointer? Please clarify. Is this still relevant?

File: src/hashmap.cpp [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    // Pre-allocate a large chunk of memory to save on all the micro-allocations
+    // that would happen if adding items one-by-one and growing as needed
+    annotations.reserve(139064548);
+    while(in.read_row(from, to, speed)){
+        add(from, to, speed);
+    }
+
+};
+
+inline void Hashmap::add(const external_nodeid_t &from, const external_nodeid_t &to, const congestion_speed_t &speed) {
```
Comment (daniel-j-h) [inline src/hashmap.cpp]: Sometimes you pass ids by const ref, sometimes you pass them by value.

Because they're integers passing by (const) value is perfectly fine - can you adapt the code everywhere?

**Comment by daniel-j-h** [comment]
Btw are we using Mason for deps? I can see the hashmap lib vendoered in third_party.

It is already bundled with mason => https://github.com/mapbox/mason/tree/master/scripts/sparsehash/2.0.2 could make sense to just use this one.

File: src/hashmap.hpp (hunk 14 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: 👍

File: src/hashmap.hpp (hunk 3 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: 👍

File: src/hashmap.hpp (hunk 13 above)
Comment (ghoshkaj) [inline src/hashmap.hpp]: This is true. However, I would need to include the Way class into every file that would want to call and use Hashmap if I did that. As Way is currently not being used anywhere else in this project, and as this is only a temporary solution, I'll opt not to do this for now.

File: src/lookup_bindings.cpp (hunk 16 above)
Comment (ghoshkaj) [inline src/lookup_bindings.cpp]: 👍

File: src/hashmap.cpp (hunk 19 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: This is the size of the wholeworld.csv file that contains all of the freeflow data currently.

File: src/hashmap.cpp (hunk 20 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: 👍

File: src/hashmap.cpp (hunk 22 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: 👍

File: src/hashmap.cpp (hunk 19 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: 👍

File: src/hashmap.cpp (hunk 21 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: I don't think this is relevant anymore.

File: src/hashmap.cpp (hunk 18 above)
Comment (ghoshkaj) [inline src/hashmap.cpp]: 👍

**Comment by ghoshkaj** [comment]
@daniel-j-h we _are_ using Mason, but we are actually using https://github.com/greg7mdp/sparsepp, not sparsehash. And sparsepp not vendored in Mason yet. We had a conversation with @springmeyer and decided not to bring it into Mason during the logistics sprint because of time constraints.

**Comment by springmeyer** [comment]
@ghoshkaj please drop a time on my calendar for getting sparsepp packaged - we can do a screenshare and it will likely take 15 min. Make sure to run through https://github.com/mapbox/mason/#trigger first.

File: test/load.test.js [hunk 23]
```diff
@@ -10,11 +11,46 @@ test('load extract', function(t) {
...
+
+test('load CSV', function(t) {
+  segmentmap.loadCSV(path.join(__dirname,'congestion/fixtures/congestion.csv'), (err) => {
+    if (err) throw err;
+    t.end();
+  });
+});
+
+test('lookup node pair', function(t) {
+  segmentmap.GetAnnotations([86909066,86909064,86909066,999], (err, resp)=> {
```
Comment (bsudekum) [inline test/load.test.js]: `GetAnnotations` should probably be `getAnnotations`.
